odl/deform/linearized_deform.py

```diff
@@ -0,0 +1,497 @@
+
+        Parameters
+        ----------
+        vector_field: `ProductSpaceVector`
+            The evaluation point, i.e. a displacement field, for the
+            deformed gradient of the template.
+        """
+        grad = odl.Gradient(self._range)
+        grad_template = grad(self.template)
+        def_grad = deform_grad(grad_template, self.displacement)
+
+        innerProdOp = odl.PointwiseInner(self.displacement.space, def_grad)
+
+        return innerProdOp(vector_field)
+
+    @property
+    def adjoint(self):
+        """Adjoint of the derivative operator.
+
+        Returns
+        -------
+        adj_op : `Operator`
+            The adjoint of the operator.
+        """
+        adj_op = LinDeforFixedTempDerivAdj(self.template, self.displacement)
+        return adj_op
+
+
+class LinDeforFixedTempDerivAdj(Operator):
+    """Adjoint operator of the derivative operator .
+
+    This operator computes the adjoint of the Gateaux derivative of the fixed
+    template linearized deformation operator for the fixed template ``I``
+    at given displacement ``v``:
+
+        J --> grad(I)(Id + v) J
+
+    where the ``J`` is the given template and
+    the ``Id`` is the identity mapping:
+
+        y --> y
+
+    and the vector field ``v`` is a displacement mapping:
+
+        y --> v(y)
+
+    Here, the ``y`` is an element in the domain of the target.
+    In diffeomorphic sense, the domain of ``I`` is the same
+    as the target.
+    """
+    def __init__(self, template, displacement):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector`
+            Fixed template deformed by the vector field.
+        displacement: `ProductSpaceVector`
+            The point that the adjoint of the Gateaux derivative needs
+            to be computed at.
+
+        Examples
+        --------
+        >>> import odl
+        >>> space = odl.uniform_discr(0, 1, 5)
+        >>> template = space.element([0, 0, 1, 0, 0])
+        >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+        >>> disp_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+        >>> vector_field = disp_field_space.element([[1, 1, 1, 2, 1]])
+        >>> op = LinDeforFixedTempDeriv(template, disp_field)
+        >>> op(vector_field)
+        uniform_discr(0.0, 1.0, 5).element([0.0, 5.0, -5.0, -10.0, 0.0])
+        """
+        assert displacement[0] in template.space
+
+        Operator.__init__(self, template.space, displacement.space,
+                          linear=True)
+
+        self.template = template
+        self.displacement = displacement
+
+    def _call(self, func):
+        """Implement ``self(func)```.
+
+        Parameters
+        ----------
+        func : `DiscreteLpVector`
+            The evaluation point, i.e. an element in the template space,
+            for the deformed gradient of the template.
+        """
+        grad = odl.Gradient(self._domain)
```

Comment:
This will force users to use the default of the `Gradient` operator (namely, forward differences). This should perhaps be customizable in some way.